services/src/main/java/org/keycloak/protocol/oid4vc/issuance/OID4VCIssuerEndpoint.java

```diff
@@ -318,82 +324,159 @@ public Response getCredentialOfferURIPreflight() {
+     * +----------+-----------+---------+---------+-----------------------------------------------------+
+     *
+     * @param credConfigId  A valid credential configuration id
+     * @param preAuthorized A flag whether the offer should be pre-authorized (requires targetUser)
+     * @param appClientId   The client id that the offer is authorized for
+     * @param appUserId     The user id that the offer is authorized for
+     * @param type          The response type, which can be 'uri' or 'qr-code'
+     * @param width         The width of the QR code image
+     * @param height        The height of the QR code image
+     * @see https://openid.net/specs/openid-4-verifiable-credential-issuance-1_0.html#name-credential-offer-endpoint
      */
     @GET
     @Produces({MediaType.APPLICATION_JSON, RESPONSE_TYPE_IMG_PNG})
     @Path("credential-offer-uri")
-    public Response getCredentialOfferURI(@QueryParam("credential_configuration_id") String vcId, @QueryParam("type") @DefaultValue("uri") OfferUriType type, @QueryParam("width") @DefaultValue("200") int width, @QueryParam("height") @DefaultValue("200") int height) {
+    public Response getCredentialOfferURI(
+            @QueryParam("credential_configuration_id") String credConfigId,
+            @QueryParam("pre_authorized") @DefaultValue("true") boolean preAuthorized,
+            @QueryParam("client_id") String appClientId,
+            @QueryParam("user_id") String appUserId,
+            @QueryParam("type") @DefaultValue("uri") OfferUriType type,
+            @QueryParam("width") @DefaultValue("200") int width,
+            @QueryParam("height") @DefaultValue("200") int height
+    ) {
         configureCors(true);
 
         AuthenticatedClientSessionModel clientSession = getAuthenticatedClientSession();
-        cors.allowedOrigins(session, clientSession.getClient());
+        UserModel userModel = clientSession.getUserSession().getUser();
+        ClientModel clientModel = clientSession.getClient();
+        RealmModel realmModel = clientModel.getRealm();
+
+        cors.allowedOrigins(session, clientModel);
         checkClientEnabled();
 
+        // Check required role to create a credential offer
+        //
+        var hasCredentialOfferRole = userModel.getRoleMappingsStream().anyMatch(rm -> rm.getName().equals(CREDENTIAL_OFFER_CREATE.getName()));
+        if (!hasCredentialOfferRole) {
+            var errorMessage = "Credential offer creation requires role: " + CREDENTIAL_OFFER_CREATE.getName();
+            throw new CorsErrorResponseException(cors,
+                    ErrorType.INVALID_CREDENTIAL_OFFER_REQUEST.toString(), errorMessage, Response.Status.BAD_REQUEST);
+        }
+
+        // Check whether the credential configuration exists in the issuer metadata
+        //
         Map<String, SupportedCredentialConfiguration> credentialsMap = OID4VCIssuerWellKnownProvider.getSupportedCredentials(session);
```

Comment:
since you are not using this map for anything else it would probably be much more performant to use `keycloaksession.realms().getClientScopesByAttributes(realm, Map.of(CredentialScopeModel.CONFIGURATION_ID, credConfigId), false)`